services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java

```diff
@@ -362,9 +385,62 @@ public Response export(UriInfo uriInfo, RealmModel realm, String format) {
                         }
                     });
 
-            String descriptor = SPMetadataDescriptor.getSPDescriptor(authnBinding, endpoint, endpoint,
-              wantAuthnRequestsSigned, wantAssertionsSigned, wantAssertionsEncrypted,
-              entityId, nameIDPolicyFormat, signingKeys, encryptionKeys);
+            // Prepare the metadata descriptor model
+            StringWriter sw = new StringWriter();
+            XMLStreamWriter writer = StaxUtil.getXMLStreamWriter(sw);
+            SAMLMetadataWriter metadataWriter = new SAMLMetadataWriter(writer);
+
+            EntityDescriptorType entityDescriptor = SPMetadataDescriptor.buildSPdescriptor(
+                authnBinding, authnBinding, endpoint, endpoint,
+                wantAuthnRequestsSigned, wantAssertionsSigned, wantAssertionsEncrypted,
+                entityId, nameIDPolicyFormat, signingKeys, encryptionKeys);
+
+            // Create the AttributeConsumingService
+            AttributeConsumingServiceType attributeConsumingService = new AttributeConsumingServiceType(attributeConsumingServiceIndex);
+            attributeConsumingService.setIsDefault(true);
+
+            if (attributeConsumingServiceName != null && attributeConsumingServiceName.length() > 0)
+            {
+                String currentLocale = realm.getDefaultLocale() == null ? "en": realm.getDefaultLocale();
+                LocalizedNameType attributeConsumingServiceNameElement = new LocalizedNameType(currentLocale);
+                attributeConsumingServiceNameElement.setValue(attributeConsumingServiceName);
+                attributeConsumingService.addServiceName(attributeConsumingServiceNameElement);
+            }
+
+            // Add the attribute mappers
+            for (IdentityProviderMapperModel attributeMapper: lstAttributeMappers) {
+                // Skip attributes with no name
+                String attributeName = attributeMapper.getConfig().get(UserAttributeMapper.ATTRIBUTE_NAME);
```

Comment:
This is specific code only for UserAttributeMapper. Maybe it should be part of mapper code itself. 

I have an idea but I am not sure this makes sense @hmlnarik could you confirm, whether it makes sense?

We could have some interface like `SamlMapperWithRequestedAttribute` with only method:
```
RequestedAttributeType getRequestedAttribute();
```

Then we could have something like this in the code to generate `attributeConsumingService`.

```
realm.getIdentityProviderMappersByAlias(getConfig().getAlias())
                    .stream()
                    .filter(mapper -> mapper instanceof ProtocolMapper)
                    .map(SamlMapperWithRequestedAttribute.class::cast)
                    .map(SamlMapperWithRequestedAttribute::getRequestedAttribute)
                    .forEach(attributeConsumingService::addRequestedAttribute);
```

WDYT?

Reply:
Sure, we can use an interface to retrieve the requested attribute and implement it in the relevant mappers. Let's wait on @hmlnarik feedback too.

Reply:
Having this interface extensible to other mappers would be great. In that case though, would it be useful to keep the descriptor alteration local to the mappers? This would require allowing mappers to directly alter the `entityDescriptor` rather than supporting a single use case with adding a requested attribute while in the descriptor generation. In code terms, that would roughly mean replacing the lines 399-438 with a loop over all `IdentityProviderMapperModel`s that are instances of a (say) `SamlMetadataDescriptorUpdater` interface. This interface would contain method `updateMetadata(EntityDescriptorType descriptor)` called from this place.

Reply:
@hmlnarik This sounds good, however, I realized there will be a problem with that. `EntityDescriptorType` is not accessible in `server-spi[-private]` modules. This means it won't be possible to implement this interface in custom/user-provided mappers. Is it a problem?

Reply:
Yes, it is possible to provide a custom IdP mapper. Could you elaborate why it would be problematic if it used `services` as a dependency and module import rather than e.g. `server-spi-private`? `server-spi-private` and `services` have the same level of support, and there this interface would not be part of `server-spi` as it would be a non-established feature

Reply:
You are right. It is the same.